### DISCUSSION TITLE
Some feedbacks and suggestions

### DISCUSSION THREAD

**Discussion by julesbouton** [sqlpage/SQLPage#147] [Ideas]
Hi again !
To complete my previous technical question, I have some feedback and some feature suggestions, that I give below. Hope it will be useful :

## Checkbox

It would be nice to be able to add some title/header before a set of checkbox. I see two ways of achieving this:
* Add an appropriate parameter : unfortunately "name" is not displayed and "label" is already used for the individual item name.
* Add a new form element of type "test", optionally with markdown ? It would allow to write a small paragraph of text in the middle of a form. Possible uses : as header for other components, as sections / subsections, or event to give instructions to the user.

## Select

If I understood well, there is no "checked" parameter (cf checkbox/radio) on 'select+multiple' elements: having some equivalent could be very useful. For the moment, if you want to edit an entry, the only way to display a list of pre-selected elements is to use checkbox with 'checked'. But if you have a lot of possible choices, this takes a lot of space. 

A possible alternative to pre-checked values for select would be the following extra-feature.

## Extra-feature

An additional kind of multiple selection element, different from select and checkbox, which could be _very_ useful, but which needs some coding.

I would imagine something like a search field with auto-completion (completion list would be given by the page and not by the browser), and when the user selects an item in the list, it would be displayed as a small label/bubble under the search field, with a 'remove' button/cross. Something like this (example from gitlab tag selector):

![image](https://github.com/lovasoa/SQLpage/assets/47297745/b1442e22-a01b-40f9-ae86-f769910ce34c)

There would be a boolean parameter to allow the user to add custom values (i.e. to add answers that do not exist yet in the list) or to limit the selection to previously existing values. There would also be the possibility to pass a list of _already_ checked values (just as 'checked' for checkbox)

Well, this one is much more work, but would be a very nice feature :)
Thanks again for all the work on this tool !

**Comment by lovasoa** [comment]
Hello again, and thanks for all the suggestions !

> It would allow to write a small paragraph of text in the middle of a form.

I think you can achieve what you want using the `description` row-level parameter, and optionally a `hidden` form field.

> there is no "checked" parameter (cf checkbox/radio) on 'select+multiple' elements

Yes, that's missing, I'll add it !

> a search field with auto-completion

That would indeed be useful, but it's not on my priority list. Would you be interested in implementing it ? I would be ready to merge a pr for that.

**Reply by lovasoa** [reply]
I added support for `selected` : https://github.com/lovasoa/SQLpage/commit/3bc76da10585d5592f5ebbc7894143df5a4700d8

**Comment by julesbouton** [comment]
> Yes, that's missing, I'll add it !

Nice, thanks !

> That would indeed be useful, but it's not on my priority list. Would you be interested in implementing it ? I would be ready to merge a pr for that.

Interested, for sure, but I am probably not competent enough :smile: 
Still, I had a look at this problem and I came up with some elements.

First, there is an implementation in [tabler](https://github.com/tabler/tabler/blob/dev/src/pages/_includes/ui/select.html) (not in the doc) which relies on [tom-select](https://tom-select.js.org). There is a preview here : https://preview.tabler.io/form-elements.html (see multiple states example)

I found how to replicate this in sqlpage, and I have a small (almost) working env attached: [mwe-select.tar.gz](https://github.com/lovasoa/SQLpage/files/13600523/mwe-select.tar.gz) However, I couldn't parse the results... (in the result page, I can only get the ID of the _last_ selected item, and not the full array of  values of the selected items)

I don't think I am able to do a pr myself, because modifying the handlebars file is not enough, since there is also a js script to improve and to add (but where ?). I am absolutely not a dev and I have a very experimental approach of these things ^^ At least, I have a custom component that almost works, and I can use it! If someone else wants to use it, the file is there.

**Reply by lovasoa** [reply]
Your solution looks great ! It doesn't need more code, just add square brackets to the variable name in the form to mark it as an array and it will work.

```sql
SELECT 'multisel' as type, 'fruit[]' as name
```

instead of ~~`'fruit' as name`~~
